Client/game_sa/CDummyPoolSA.cpp

```diff
@@ -13,63 +13,93 @@
 {
-    if (m_pLodBackup)
+    if (m_pOriginalElementsBackup)
         return;
 
-    m_pLodBackup = std::make_unique<std::array<CEntitySAInterface*, MAX_DUMMIES>>();
+    m_pOriginalElementsBackup = std::make_unique<pool_backup_t>();
 
-    for (int i = 0; i < MAX_DUMMIES; i++)
+    auto pDummyPool = (*m_ppDummyPoolInterface);
+    for (size_t i = 0; i < MAX_DUMMIES_DEFAULT; i++)
     {
-        CEntitySAInterface* object = (*m_ppDummyPoolInterface)->GetObject(i);
-        (*m_pLodBackup)[i] = object->m_pLod;
-        object->m_pLod = nullptr;
+        if (pDummyPool->IsContains(i))
+        {
+            CEntitySAInterface* building = pDummyPool->GetObject(i);
+
+            pGame->GetWorld()->Remove(building, CDummyPool_Destructor);
+            building->RemoveRWObjectWithReferencesCleanup();
+
+            pDummyPool->Release(i);
+
+            (*m_pOriginalElementsBackup)[i].first = true;
+            (*m_pOriginalElementsBackup)[i].second = *building;
+        }
+        else
+        {
+            (*m_pOriginalElementsBackup)[i].first = false;
+        }
     }
 }
 
-void CDummyPoolSA::RestoreAllBuildingsLods()
+void CDummyPoolSA::RestoreBackup()
 {
-    if (!m_pLodBackup)
+    if (!m_pOriginalElementsBackup)
         return;
 
-    for (int i = 0; i < MAX_DUMMIES; i++)
+    auto& originalData = *m_pOriginalElementsBackup;
+    auto  pDummyPool = (*m_ppDummyPoolInterface);
+    for (size_t i = 0; i < MAX_DUMMIES_DEFAULT; i++)
     {
-        CEntitySAInterface* object = (*m_ppDummyPoolInterface)->GetObject(i);
-        object->m_pLod = (*m_pLodBackup)[i];
+        if (originalData[i].first)
+        {
+            pDummyPool->AllocateAt(i);
+            auto pDummy = pDummyPool->GetObject(i);
+            *pDummy = originalData[i].second;
+
+            pGame->GetWorld()->Add(pDummy, CDummyPool_Constructor);
+        }
     }
 
-    m_pLodBackup.release();
+    m_pOriginalElementsBackup = nullptr;
 }
 
 void CDummyPoolSA::UpdateBuildingLods(void* oldPool, void* newPool)
 {
-    const uint32_t offset = (uint32_t)newPool - (uint32_t)oldPool;
+    const std::uint32_t offset = (std::uint32_t)newPool - (std::uint32_t)oldPool;
 
-    if (m_pLodBackup)
+    if (m_pOriginalElementsBackup)
     {
-        for (int i = 0; i < MAX_DUMMIES; i++)
+        for (int i = 0; i < MAX_DUMMIES_DEFAULT; i++)
         {
-            if ((*m_pLodBackup)[i] != nullptr)
+            if ((*m_pOriginalElementsBackup)[i].first)
             {
-                (*m_pLodBackup)[i] = (CEntitySAInterface*)((uint32_t)(*m_pLodBackup)[i] + offset);
+                CEntitySAInterface* object = &(*m_pOriginalElementsBackup)[i].second;
+                if (object->m_pLod)
+                {
+                    object->m_pLod = (CEntitySAInterface*)((std::uint32_t)(object->m_pLod) + offset);
+                }
             }
         }
     }
     else
     {
-        for (int i = 0; i < MAX_DUMMIES; i++)
+        for (int i = 0; i < MAX_DUMMIES_DEFAULT; i++)
```

Comment:
The loop is the same. Move the if statement to the loop instead for better readability

Reply:
They are not same

Reply:
> They are not the same

```cpp
for (int i = 0; i < MAX_DUMMIES_DEFAULT; i++)
{
    if (m_pOriginalElementsBackup && (*m_pOriginalElementsBackup)[i].first)
    {

    }
    else
    {
    }
```

Reply:
In your example `(*m_pOriginalElementsBackup)[i].first == false` causes using second loop.

Reply:
Thats your code but moved the `if` statement to the loop

Reply:
No. Your code is bugged. Negative `(*m_pOriginalElementsBackup)[i].first` causes using `else` branch.
Read it again

Reply:
> No. Your code is bugged. Negative `(*m_pOriginalElementsBackup)[i].first` causes using `else` branch. Read it again

thats literally your code.
```cpp
if (m_pOriginalElementsBackup)
{
    for (int i = 0; i < MAX_DUMMIES_DEFAULT; i++)
    {
        if ((*m_pOriginalElementsBackup)[i].first)
        {
            CEntitySAInterface* object = &(*m_pOriginalElementsBackup)[i].second;
            if (object->m_pLod)
            {
                object->m_pLod = (CEntitySAInterface*)((std::uint32_t)(object->m_pLod) + offset);
            }
        }
    }
}
else
{
    for (int i = 0; i < MAX_DUMMIES_DEFAULT; i++)
    {
        CEntitySAInterface* object = (*m_ppDummyPoolInterface)->GetObject(i);
        if (object->m_pLod)
        {
            object->m_pLod = (CEntitySAInterface*)((std::uint32_t)object->m_pLod + offset);
        }
    }
}
```
When you move the if statement to the loop, you will get the code above (and below):
```cpp
for (int i = 0; i < MAX_DUMMIES_DEFAULT; i++)
{
    if(m_pOriginalElementsBackup && (*m_pOriginalElementsBackup)[i].first)
    {

    }
    else
    {

    }
}
```
Negative `.first` have no effect on the `if` statement

Reply:
You could even shorten it to this:
```cpp
for (int i = 0; i < MAX_DUMMIES_DEFAULT; i++)
{
    CEntitySAInterface* object;

    if(m_pOriginalElementsBackup && (*m_pOriginalElementsBackup)[i].first)
    {
        object = &(*m_pOriginalElementsBackup)[i].second;
    }
    else
    {
        object = (*m_ppDummyPoolInterface)->GetObject(i);
    }
    if (object->m_pLod)
    {
        object->m_pLod = (CEntitySAInterface*)((std::uint32_t)object->m_pLod + offset);
    }
}
```

Reply:
it has, when `m_pOriginalElementsBackup` is not null and `(*m_pOriginalElementsBackup)[i].first` is false.
By your logic loop will use `(*m_ppDummyPoolInterface)->GetObject(i);`

Reply:
again, you can detect that and have 1 loop instead of 2 with nearly the same content.

Reply:
> again, you can detect that and have 1 loop instead of 2 with nearly the same content.

Please, be totally explicit. I can't read your mind. You twice suggested me totally wrong code and now you're writing about `else if` branch.

Reply:
> Please, be totally explicit. I can't read your mind. You twice suggested me totally wrong code and now you're writing about `else if` branch.

It was always about `else`. You have `else` in your code.
Just put it in the loop itself.
```cpp
for (int i = 0; i < MAX_DUMMIES_DEFAULT; i++)
{
    CEntitySAInterface* object;

    if(m_pOriginalElementsBackup)
    {
        if(!(*m_pOriginalElementsBackup)[i].first)
            continue;
        object = &(*m_pOriginalElementsBackup)[i].second;
    }
    else
    {
        object = (*m_ppDummyPoolInterface)->GetObject(i);
    }

    if (!object->m_pLod)
        continue;

    object->m_pLod = (CEntitySAInterface*)((std::uint32_t)object->m_pLod + offset);
}
```

Reply:
> It was always about else. You have else in your code.

You didn't provide much context about what do you want to see.
Horewer, it's spaghetti. I'd prefer to extract method instead.

```c++
if (m_pOriginalElementsBackup)
    UpdateLodPointersInBackup(offset);
else
    UpdateLodPointersInPool(offset);
```

Reply:
> You didn't provide much context about what do you want to see.

🤨

> Horewer, it's spaghetti. I'd prefer to extract method instead.

that could work too but in this case it would be better for simple if statement imo

Reply:
No, your example is unnecessary complex. Two additional continue in one loop in different levels are confusing

Reply:
Having two near identical for loops is confusing.
This example reduces the code. Two `continue` isnt confusing.
This example is straightforward.
We have much more confusing code in our codebase thanks to these huge structures